### ISSUE TITLE
_msearch requires newline

### ISSUE THREAD

**Issue by kananindzya** [elastic/elasticsearch-java#41]
Hi,
1. MsearchTemplateRequest
I am trying to write tests with latest version of elasticsearch-java(7.15.1), elasticsearch(docker.elastic.co/elasticsearch/elasticsearch:7.15.0) for elastic-apm-agent. And got exception message below:
```
The msearch request must be terminated by a newline [\n]
```
As I understand it, the problem is in the serialization of the request - instead of json separated new line, the client sends an array of objects

Steps to reproduce:
```
 IndexResponse ir = doIndex(new IndexRequest(new IndexRequest.Builder()
            .index(INDEX)
            .type(DOC_TYPE)
            .id(DOC_ID)
            .refresh(JsonValue.TRUE)
            .document(Map.of(FOO, BAR))));
 PutScriptRequest putScriptRequest = new PutScriptRequest(builder -> builder
            .id("elastic-search-template")
            .script(new StoredScript(scriptBuilder ->
                scriptBuilder.lang(ScriptLanguage.Mustache)
                    .source("{" +
                        "  \"query\": { \"term\" : { \"{{field}}\" : \"{{value}}\" } }," +
                        "  \"size\" : \"{{size}}\"" +
                        "}")
            )));

        client.putScript(putScriptRequest);

   MsearchTemplateRequest multiRequest = new MsearchTemplateRequest(builder -> builder.searchTemplates(
            new TemplateItem(itemBuilder -> itemBuilder.index(INDEX)
                .id("elastic-search-template")
                .params(Map.of("field", JsonData.of(FOO), "value", JsonData.of(BAR), "size", JsonData.of(5))))
        ));

        try {
            MsearchTemplateResponse<Map> response = doMultiSearchTemplate(multiRequest, Map.class);
        } catch (ApiException e) {
            // reason -> {JsonStringImpl@7318} ""The msearch request must be terminated by a newline [\n]""
            LOGGER.error("Error", e);
        }
```
Addition logs:
```
2183 [main] DEBUG org.apache.http.impl.nio.conn.ManagedNHttpClientConnectionImpl  - http-outgoing-0 127.0.0.1:59098<->127.0.0.1:49297[ACTIVE][rw:r]: Event set [w]
2183 [I/O dispatcher 1] DEBUG org.apache.http.impl.nio.client.InternalIODispatch  - http-outgoing-0 [ACTIVE] Request ready
2183 [I/O dispatcher 1] DEBUG org.apache.http.impl.nio.client.MainClientExec  - [exchange: 4] Attempt 1 to execute request
2183 [I/O dispatcher 1] DEBUG org.apache.http.impl.nio.client.MainClientExec  - [exchange: 4] Target auth state: UNCHALLENGED
2183 [I/O dispatcher 1] DEBUG org.apache.http.impl.nio.client.MainClientExec  - [exchange: 4] Proxy auth state: UNCHALLENGED
2183 [I/O dispatcher 1] DEBUG org.apache.http.impl.nio.conn.ManagedNHttpClientConnectionImpl  - http-outgoing-0 127.0.0.1:59098<->127.0.0.1:49297[ACTIVE][rw:w]: Set timeout 30000
2183 [I/O dispatcher 1] DEBUG org.apache.http.headers  - http-outgoing-0 >> POST /_msearch/template HTTP/1.1
2183 [I/O dispatcher 1] DEBUG org.apache.http.headers  - http-outgoing-0 >> Content-Length: 103
2183 [I/O dispatcher 1] DEBUG org.apache.http.headers  - http-outgoing-0 >> Content-Type: application/json; charset=UTF-8
2183 [I/O dispatcher 1] DEBUG org.apache.http.headers  - http-outgoing-0 >> Host: localhost:49297
2183 [I/O dispatcher 1] DEBUG org.apache.http.headers  - http-outgoing-0 >> Connection: Keep-Alive
2183 [I/O dispatcher 1] DEBUG org.apache.http.headers  - http-outgoing-0 >> User-Agent: elasticsearch-java/7.15.0-SNAPSHOT (Java/11.0.11)
2183 [I/O dispatcher 1] DEBUG org.apache.http.headers  - http-outgoing-0 >> X-Elastic-Client-Meta: es=7.15.0-SNAPSHOT,jv=11,t=7.15.0-SNAPSHOT,hc=4.1.4
2183 [I/O dispatcher 1] DEBUG org.apache.http.headers  - http-outgoing-0 >> Authorization: Basic ZWxhc3RpYy11c2VyOmVsYXN0aWMtcGFzcw==
2184 [I/O dispatcher 1] DEBUG org.apache.http.impl.nio.conn.ManagedNHttpClientConnectionImpl  - http-outgoing-0 127.0.0.1:59098<->127.0.0.1:49297[ACTIVE][rw:w]: Event set [w]
2184 [I/O dispatcher 1] DEBUG org.apache.http.impl.nio.client.InternalIODispatch  - http-outgoing-0 [ACTIVE] Output ready
2184 [I/O dispatcher 1] DEBUG org.apache.http.impl.nio.client.MainClientExec  - [exchange: 4] produce content
2184 [I/O dispatcher 1] DEBUG org.apache.http.impl.nio.client.MainClientExec  - [exchange: 4] Request completed
2184 [I/O dispatcher 1] DEBUG org.apache.http.impl.nio.client.InternalIODispatch  - http-outgoing-0 [ACTIVE] [content length: 103; pos: 103; completed: true]
2184 [I/O dispatcher 1] DEBUG org.apache.http.impl.nio.conn.ManagedNHttpClientConnectionImpl  - http-outgoing-0 127.0.0.1:59098<->127.0.0.1:49297[ACTIVE][rw:w]: 452 bytes written
2184 [I/O dispatcher 1] DEBUG org.apache.http.wire  - http-outgoing-0 >> "POST /_msearch/template HTTP/1.1[\r][\n]"
2184 [I/O dispatcher 1] DEBUG org.apache.http.wire  - http-outgoing-0 >> "Content-Length: 103[\r][\n]"
2184 [I/O dispatcher 1] DEBUG org.apache.http.wire  - http-outgoing-0 >> "Content-Type: application/json; charset=UTF-8[\r][\n]"
2184 [I/O dispatcher 1] DEBUG org.apache.http.wire  - http-outgoing-0 >> "Host: localhost:49297[\r][\n]"
2184 [I/O dispatcher 1] DEBUG org.apache.http.wire  - http-outgoing-0 >> "Connection: Keep-Alive[\r][\n]"
2184 [I/O dispatcher 1] DEBUG org.apache.http.wire  - http-outgoing-0 >> "User-Agent: elasticsearch-java/7.15.0-SNAPSHOT (Java/11.0.11)[\r][\n]"
2184 [I/O dispatcher 1] DEBUG org.apache.http.wire  - http-outgoing-0 >> "X-Elastic-Client-Meta: es=7.15.0-SNAPSHOT,jv=11,t=7.15.0-SNAPSHOT,hc=4.1.4[\r][\n]"
2184 [I/O dispatcher 1] DEBUG org.apache.http.wire  - http-outgoing-0 >> "Authorization: Basic ZWxhc3RpYy11c2VyOmVsYXN0aWMtcGFzcw==[\r][\n]"
2184 [I/O dispatcher 1] DEBUG org.apache.http.wire  - http-outgoing-0 >> "[\r][\n]"
2184 [I/O dispatcher 1] DEBUG org.apache.http.wire  - http-outgoing-0 >> "[{"id":"elastic-search-template","index":["my-index"],"params":{"size":5,"field":"foo","value":"bar"}}]"
2185 [I/O dispatcher 1] DEBUG org.apache.http.impl.nio.client.InternalIODispatch  - http-outgoing-0 [ACTIVE] Request ready
2185 [I/O dispatcher 1] DEBUG org.apache.http.impl.nio.conn.ManagedNHttpClientConnectionImpl  - http-outgoing-0 127.0.0.1:59098<->127.0.0.1:49297[ACTIVE][r:w]: Event cleared [w]
2188 [I/O dispatcher 1] DEBUG org.apache.http.impl.nio.conn.ManagedNHttpClientConnectionImpl  - http-outgoing-0 127.0.0.1:59098<->127.0.0.1:49297[ACTIVE][r:r]: 696 bytes read
2188 [I/O dispatcher 1] DEBUG org.apache.http.wire  - http-outgoing-0 << "HTTP/1.1 400 Bad Request[\r][\n]"
2188 [I/O dispatcher 1] DEBUG org.apache.http.wire  - http-outgoing-0 << "X-elastic-product: Elasticsearch[\r][\n]"
2188 [I/O dispatcher 1] DEBUG org.apache.http.wire  - http-outgoing-0 << "Warning: 299 Elasticsearch-7.15.0-79d65f6e357953a5b3cbcc5e2c7c21073d89aa29 "Elasticsearch built-in security features are not enabled. Without authentication, your cluster could be accessible to anyone. See https://www.elastic.co/guide/en/elasticsearch/reference/7.15/security-minimal-setup.html to enable security."[\r][\n]"
2188 [I/O dispatcher 1] DEBUG org.apache.http.wire  - http-outgoing-0 << "content-type: application/json; charset=UTF-8[\r][\n]"
2188 [I/O dispatcher 1] DEBUG org.apache.http.wire  - http-outgoing-0 << "content-length: 249[\r][\n]"
2188 [I/O dispatcher 1] DEBUG org.apache.http.wire  - http-outgoing-0 << "[\r][\n]"
2188 [I/O dispatcher 1] DEBUG org.apache.http.wire  - http-outgoing-0 << "{"error":{"root_cause":[{"type":"illegal_argument_exception","reason":"The msearch request must be terminated by a newline [\n]"}],"type":"illegal_argument_exception","reason":"The msearch request must be terminated by a newline [\n]"},"status":400}"
2188 [I/O dispatcher 1] DEBUG org.apache.http.headers  - http-outgoing-0 << HTTP/1.1 400 Bad Request
2188 [I/O dispatcher 1] DEBUG org.apache.http.headers  - http-outgoing-0 << X-elastic-product: Elasticsearch
2189 [I/O dispatcher 1] DEBUG org.apache.http.headers  - http-outgoing-0 << Warning: 299 Elasticsearch-7.15.0-79d65f6e357953a5b3cbcc5e2c7c21073d89aa29 "Elasticsearch built-in security features are not enabled. Without authentication, your cluster could be accessible to anyone. See https://www.elastic.co/guide/en/elasticsearch/reference/7.15/security-minimal-setup.html to enable security."
2189 [I/O dispatcher 1] DEBUG org.apache.http.headers  - http-outgoing-0 << content-type: application/json; charset=UTF-8
2189 [I/O dispatcher 1] DEBUG org.apache.http.headers  - http-outgoing-0 << content-length: 249
2189 [I/O dispatcher 1] DEBUG org.apache.http.impl.nio.client.InternalIODispatch  - http-outgoing-0 [ACTIVE(249)] Response received
2189 [I/O dispatcher 1] DEBUG org.apache.http.impl.nio.client.MainClientExec  - [exchange: 4] Response received HTTP/1.1 400 Bad Request
2189 [I/O dispatcher 1] DEBUG org.apache.http.impl.nio.client.InternalIODispatch  - http-outgoing-0 [ACTIVE(249)] Input ready
2189 [I/O dispatcher 1] DEBUG org.apache.http.impl.nio.client.MainClientExec  - [exchange: 4] Consume content
2189 [I/O dispatcher 1] DEBUG org.apache.http.impl.nio.client.InternalHttpAsyncClient  - [exchange: 4] Connection can be kept alive indefinitely
2189 [I/O dispatcher 1] DEBUG org.apache.http.impl.nio.client.MainClientExec  - [exchange: 4] Response processed
2189 [I/O dispatcher 1] DEBUG org.apache.http.impl.nio.client.InternalHttpAsyncClient  - [exchange: 4] releasing connection
2189 [I/O dispatcher 1] DEBUG org.apache.http.impl.nio.conn.ManagedNHttpClientConnectionImpl  - http-outgoing-0 127.0.0.1:59098<->127.0.0.1:49297[ACTIVE][r:r]: Remove attribute http.nio.exchange-handler
2190 [I/O dispatcher 1] DEBUG org.apache.http.impl.nio.conn.PoolingNHttpClientConnectionManager  - Releasing connection: [id: http-outgoing-0][route: {}->http://localhost:49297][total kept alive: 0; route allocated: 1 of 10; total allocated: 1 of 30]
2190 [I/O dispatcher 1] DEBUG org.apache.http.impl.nio.conn.PoolingNHttpClientConnectionManager  - Connection [id: http-outgoing-0][route: {}->http://localhost:49297] can be kept alive indefinitely
```
What am I doing incorrectly?

2. Same thing for multi-search request.
```
 MsearchRequest multiSearchRequest = new MsearchRequest(builder -> builder
            .index(INDEX)
            .searches(
                Json.createObjectBuilder()
                    .add("index", Json.createArrayBuilder().add(INDEX))
                    .add("types", Json.createArrayBuilder())
                    .add("search_type", "query_then_fetch")
                    .add("ccs_minimize_roundtrips", true)
                    .build(),
                Json.createObjectBuilder()
                    .add("query", Json.createObjectBuilder()
                        .add("match", Json.createObjectBuilder()
                            .add(FOO, Json.createObjectBuilder()
                                .add("query", BAR)
                                .add("operator", "OR")))
                        .build()).build())
            .typedKeys(true));

        ApiException apiException = null;
        try {
            MsearchResponse response = doMultiSearch(multiSearchRequest, Map.class);
        } catch (ApiException e) {
            apiException = e;
        }
```

```
1836 [I/O dispatcher 1] DEBUG org.apache.http.wire  - http-outgoing-0 >> "POST /my-index/_msearch?typed_keys=true HTTP/1.1[\r][\n]"
1836 [I/O dispatcher 1] DEBUG org.apache.http.wire  - http-outgoing-0 >> "Content-Length: 159[\r][\n]"
1836 [I/O dispatcher 1] DEBUG org.apache.http.wire  - http-outgoing-0 >> "Content-Type: application/json; charset=UTF-8[\r][\n]"
1836 [I/O dispatcher 1] DEBUG org.apache.http.wire  - http-outgoing-0 >> "Host: localhost:49173[\r][\n]"
1836 [I/O dispatcher 1] DEBUG org.apache.http.wire  - http-outgoing-0 >> "Connection: Keep-Alive[\r][\n]"
1836 [I/O dispatcher 1] DEBUG org.apache.http.wire  - http-outgoing-0 >> "User-Agent: elasticsearch-java/7.15.0-SNAPSHOT (Java/11.0.11)[\r][\n]"
1836 [I/O dispatcher 1] DEBUG org.apache.http.wire  - http-outgoing-0 >> "X-Elastic-Client-Meta: es=7.15.0-SNAPSHOT,jv=11,t=7.15.0-SNAPSHOT,hc=4.1.4[\r][\n]"
1836 [I/O dispatcher 1] DEBUG org.apache.http.wire  - http-outgoing-0 >> "Authorization: Basic ZWxhc3RpYy11c2VyOmVsYXN0aWMtcGFzcw==[\r][\n]"
1836 [I/O dispatcher 1] DEBUG org.apache.http.wire  - http-outgoing-0 >> "[\r][\n]"
1836 [I/O dispatcher 1] DEBUG org.apache.http.wire  - http-outgoing-0 >> "[{"index":["my-index"],"types":[],"search_type":"query_then_fetch","ccs_minimize_roundtrips":true},{"query":{"match":{"foo":{"query":"bar","operator":"OR"}}}}]"
1836 [I/O dispatcher 1] DEBUG org.apache.http.impl.nio.client.InternalIODispatch  - http-outgoing-0 [ACTIVE] Request ready
1836 [I/O dispatcher 1] DEBUG org.apache.http.impl.nio.conn.ManagedNHttpClientConnectionImpl  - http-outgoing-0 127.0.0.1:60408<->127.0.0.1:49173[ACTIVE][r:w]: Event cleared [w]
1840 [I/O dispatcher 1] DEBUG org.apache.http.impl.nio.conn.ManagedNHttpClientConnectionImpl  - http-outgoing-0 127.0.0.1:60408<->127.0.0.1:49173[ACTIVE][r:r]: 696 bytes read
1840 [I/O dispatcher 1] DEBUG org.apache.http.wire  - http-outgoing-0 << "HTTP/1.1 400 Bad Request[\r][\n]"
1840 [I/O dispatcher 1] DEBUG org.apache.http.wire  - http-outgoing-0 << "X-elastic-product: Elasticsearch[\r][\n]"
1840 [I/O dispatcher 1] DEBUG org.apache.http.wire  - http-outgoing-0 << "Warning: 299 Elasticsearch-7.15.0-79d65f6e357953a5b3cbcc5e2c7c21073d89aa29 "Elasticsearch built-in security features are not enabled. Without authentication, your cluster could be accessible to anyone. See https://www.elastic.co/guide/en/elasticsearch/reference/7.15/security-minimal-setup.html to enable security."[\r][\n]"
1840 [I/O dispatcher 1] DEBUG org.apache.http.wire  - http-outgoing-0 << "content-type: application/json; charset=UTF-8[\r][\n]"
1840 [I/O dispatcher 1] DEBUG org.apache.http.wire  - http-outgoing-0 << "content-length: 249[\r][\n]"
1840 [I/O dispatcher 1] DEBUG org.apache.http.wire  - http-outgoing-0 << "[\r][\n]"
1840 [I/O dispatcher 1] DEBUG org.apache.http.wire  - http-outgoing-0 << "{"error":{"root_cause":[{"type":"illegal_argument_exception","reason":"The msearch request must be terminated by a newline [\n]"}],"type":"illegal_argument_exception","reason":"The msearch request must be terminated by a newline [\n]"},"status":400}"
1840 [I/O dispatcher 1] DEBUG org.apache.http.headers  - http-outgoing-0 << HTTP/1.1 400 Bad Request
1840 [I/O dispatcher 1] DEBUG org.apache.http.headers  - http-outgoing-0 << X-elastic-product: Elasticsearch
1840 [I/O dispatcher 1] DEBUG org.apache.http.headers  - http-outgoing-0 << Warning: 299 Elasticsearch-7.15.0-79d65f6e357953a5b3cbcc5e2c7c21073d89aa29 "Elasticsearch built-in security features are not enabled. Without authentication, your cluster could be accessible to anyone. See https://www.elastic.co/guide/en/elasticsearch/reference/7.15/security-minimal-setup.html to enable security."
```
3. It would be great if you add api that allow convert Query to JsonValue.
Does current api allow convert `TermQuery` or `MatchQuery` to `JsonValue`?
```
.searches(Json.createArrayBuilder().add(Json.createObjectBuilder()
                .add("query", Json.createObjectBuilder().add("match", Json.createObjectBuilder().add(FOO, BAR)))
                .build()).build()))
```

**Comment by swallez** [comment]
Hi @kananindzya. Sorry for the late answer. Msearch is not entirely implemented in the new Java Client version 7.15. This will be fixed in version 7.16 that will be released soon, that will also avoid manual creation of JSON structures.

**Comment by swallez** [comment]
mulitisearch requests and their unusual nd-json encoding have been implemented [and tested](https://github.com/elastic/elasticsearch-java/commit/3c083e5e314fc72f691495b3ead50a0e8173603d). This will be part of 7.16.
